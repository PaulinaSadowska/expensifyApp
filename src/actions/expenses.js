import { v1 as uuid } from 'uuid'
import moment from 'moment';
import database from '../firebase/firebase'

// ADD_EXPENSE
export const addExpense = (expense = {}) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            note = "",
            amount = 0,
            createdAt = moment().valueOf()
        } = expenseData;
        const expense = { description, note, amount, createdAt }
        database.collection("expenses").add(expense).then((ref) => {
            console.log("saved", ref)
            dispatch(addExpense({
                id: ref.id,
                ...expense
            }))
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});