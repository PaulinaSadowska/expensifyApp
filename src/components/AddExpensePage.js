import React from 'react';

export const AddExpensePage = () => (
    <p>add expense page</p>
);