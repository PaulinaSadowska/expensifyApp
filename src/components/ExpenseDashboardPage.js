import React from 'react';

export const ExpenseDashboardPage = () => (
    <p>expense dashboard page</p>
);