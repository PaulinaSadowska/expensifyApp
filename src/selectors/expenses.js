import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        const createdMoment = moment(expense.createdAt);
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdMoment, "day") : true;
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        return startDateMatch && endDateMatch && textMatch;
    }).sort((expense1, expense2) => {
        if(sortBy == "amount"){
            return (expense1.amount < expense2.amount) ? 1 : -1;
        } else if(sortBy == "date"){
            return (expense1.createdAt < expense2.createdAt) ? 1 : -1;
        }
    })
};