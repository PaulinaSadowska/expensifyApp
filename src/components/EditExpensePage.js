import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {

    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense)
        this.props.history.push("/")
    }

    onRemoveClick = () => {
        this.props.removeExpense({ id: this.props.expense.id })
        this.props.history.push("/")
    }

    render() {
        return <div>
            Editing the expense: {this.props.expense.description}
            <div><button onClick={this.props.history.goBack}>back</button></div>
            <ExpenseForm
                expense={this.props.expense}
                onSubmit={this.onSubmit}
            />
            <button onClick={this.onRemoveClick}>Remove</button>
        </div>
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((exp) => exp.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => ({
    removeExpense: (data) => dispatch(removeExpense(data)),
    editExpense: (id, expense) => dispatch(editExpense(id, expense))
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)