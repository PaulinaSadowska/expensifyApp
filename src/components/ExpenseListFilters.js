import React from 'react';
import { DateRangePicker } from 'react-dates';
import { connect } from 'react-redux'
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from '../actions/filters'

export class ExpenseListFilters extends React.Component {

    state = {
        calendarFocused: undefined
    }

    onFocusChange = (focused) => {
        this.setState(() => ({
            calendarFocused: focused
        }))
    }

    onDatesChanged = ({ startDate, endDate }) => {
        console.log(startDate)
        this.props.setStartDate(startDate)
        this.props.setEndDate(endDate)
    }

    onTextFilterChanged = (e) => {
        this.props.setTextFilter(e.target.value)
    }

    onSortByChanged = (e) => {
        if (e.target.value == "amount") {
            this.props.sortByAmount()
        } else if (e.target.value == "date") {
            this.props.sortByDate()
        }
    }

    render() {
        return <div className="content-container">
            <div className="input-group">
                <div className="input-group__item">
                    <input
                        className="text-input"
                        type="text"
                        placeholder="Search expenses"
                        value={this.props.filters.text}
                        onChange={this.onTextFilterChanged} />
                </div>
                <div className="input-group__item">
                    <select
                        className="select"
                        value={this.props.filters.sortBy}
                        onChange={this.onSortByChanged}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                </div>
                <div className="input-group__item">
                    <DateRangePicker
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChanged}
                        focusedInput={this.state.calendarFocused}
                        startDateId="startId1"
                        endDateId="endId1"
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={2}
                        isOutsideRange={() => false}
                        showClearDates={true} />
                </div>
            </div>
        </div>
    }
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (textFilter) => dispatch(setTextFilter(textFilter)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate()),
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);