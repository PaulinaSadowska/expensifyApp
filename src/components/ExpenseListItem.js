import React from 'react';
import { connect } from 'react-redux';
import { removeExpense } from '../actions/expenses'
import {Link} from 'react-router-dom'

const ExpenseListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>${(amount / 100).toFixed(2)} - {createdAt}</p>
        <button onClick={() => {
            dispatch(removeExpense({ id }))
        }}>Remove</button>
    </div>
);

export default connect()(ExpenseListItem)