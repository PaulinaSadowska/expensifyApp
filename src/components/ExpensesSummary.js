import React from 'react'
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';
import numeral from 'numeral';
import expenses from '../selectors/expenses';

export const ExpensesSummary = (props) => {
    const { expenses } = props
    const totalPriceFormatted = numeral(getExpensesTotal(expenses)/100).format('$0,0.00')
    const expenseWord = (expenses.length===1) ? "expense" : "expenses"
    return <p>
        <h2>Viewing {expenses.length} {expenseWord} totalling {totalPriceFormatted}</h2>
    </p>  
};

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpensesSummary);