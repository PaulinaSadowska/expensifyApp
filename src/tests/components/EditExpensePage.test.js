import { EditExpensePage } from '../../components/EditExpensePage';
import React from 'react';
import { shallow } from 'enzyme';
import { expenses } from '../fixtures/expenses'

let removeExpense, editExpense, history, wrapper;

beforeEach(() => {
    removeExpense = jest.fn()
    editExpense = jest.fn()
    history = { push: jest.fn() }
    wrapper = shallow(
        <EditExpensePage
            removeExpense={removeExpense}
            editExpense={editExpense}
            history={history}
            expense={expenses[2]} />
    );
})

test('should render edit expense page', () => {
    expect(wrapper).toMatchSnapshot();
})

test('should handle edit expense', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[2])
    expect(editExpense).toHaveBeenCalledWith(expenses[2].id, expenses[2])
    expect(history.push).toHaveBeenCalledWith("/")
})

test('should handle remove expense', () => {
    wrapper.find('button').simulate('click')
    expect(removeExpense).toHaveBeenCalledWith({ id: expenses[2].id })
    expect(history.push).toHaveBeenCalledWith("/")
})