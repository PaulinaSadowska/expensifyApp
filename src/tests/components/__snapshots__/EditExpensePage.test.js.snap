// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render edit expense page 1`] = `
<div>
  Editing the expense: 
  Credit card
  <div>
    <button>
      back
    </button>
  </div>
  <ExpenseForm
    expense={
      Object {
        "amount": 4500,
        "createdAt": 1135810800000,
        "description": "Credit card",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
